### DISCUSSION TITLE
Some questions about configuration of LSPs

### DISCUSSION THREAD

**Discussion by zhang-stephen** [mason-org/mason.nvim#57] [Q&A] [answered]
I am considering that migration from nvim-lsp-installer, and I still have some questions about this plugin.

Does `mason-lspconfig` support to configure servers installed by package manager, e.g. `ccls` or `clangd`?

And I noticed that [there](https://github.com/williamboman/mason.nvim/blob/9399b0e66f7cffeb88fe112546ebd06c57f46ed2/lua/mason-lspconfig/init.lua#L112) is a default handler when LSP is configured. How could I pass a user-defined configuration of a LSP server to nvim-lspconfig by lua table? I think create handler for each server is not convenient.

another point is, the package name in registry, for example, I want to install `bashls`, which command is correct?

+ `MasonInstall bashls`
+ `MasonInstall bash-language-server`

I think it would be confused if the package name(for LSP only) is different with nvim-lspconfig or nvim-lsp-installer...

**Comment by williamboman** [comment, marked as answer]
Hello!

> Does mason-lspconfig support to configure servers installed by package manager, e.g. ccls or clangd?

`mason-lspconfig` is not explicitly responsible for setting up servers, it only registers hooks with `lspconfig` as well as provide some other APIs that makes it a bit more seamless - for example an `:LspInstall` command, and an API that allow you to "automatically" set up the servers installed via mason so you don't have to do so explicitly for each server in your config.

For servers installed outside of mason (e.g. system package manager) you should just directly interact with lspconfig as mason is not involved in that at all.

> And I noticed that [there](https://github.com/williamboman/mason.nvim/blob/9399b0e66f7cffeb88fe112546ebd06c57f46ed2/lua/mason-lspconfig/init.lua#L112) is a default handler when LSP is configured. How could I pass a user-defined configuration of a LSP server to nvim-lspconfig by lua table? I think create handler for each server is not convenient.

I need to write docs on what these things do - so this function is not called by mason itself, it's an opt-in functionality for people who prefer to use mason as the main control interface to set up servers, for example:

```lua
local lspconfig = require("lspconfig")

require("mason-lspconfig").setup_handlers {
    -- This is a default handler that will be called for each installed server (also for new servers that are installed during a session)
    function (server_name)
       lspconfig[server.name].setup {}
   end,
   -- You can also override the default handler for specific servers by providing them as keys, like so:
   ["rust_analyzer"] = function ()
     require("rust-tools").setup {}
   end
}
```

> another point is, the package name in registry, for example, I want to install bashls, which command is correct?

`bashls` is lspconfig's own name for `bash-language-server`. Mason uses the canonical name of each package, so `:MasonInstall bash-language-server` is the correct option. Note that if you also decide to use `mason-lspconfig`, there will be a `:LspInstall` command available which accepts the lspconfig name, like `:LspInstall bashls` (this is primarily done to enable a smooth transition from https://github.com/williamboman/nvim-lsp-installer).

> I think it would be confused if the package name(for LSP only) is different with nvim-lspconfig or nvim-lsp-installer...

I understand, however, lspconfig is an optional plugin that comes with a set of unsupported, best-effort, server configurations. You can use Neovim's builtin LSP client perfectly fine without `lspconfig`, and from what I can see in terms of work being done in Neovim core, more and more functionality from lspconfig is being introduced in Neovim core.

**Reply by zhang-stephen** [reply]
Cool! Looking forward the complete documents.

and there is a few language servers supported by the hook for 'automatically' configuring in my opinion, isn't it?
